### PR TITLE
cosmos lazy decryption

### PR DESCRIPTION
**Pull request by moderakh**
This PR Provides EncryptableItem<T> and DecryptableItem<T> as wrapper on T to provide a way for lazy decryption.

The PR captures the APIs and design which discussed earlier and implemented on the dot-net side:
https://github.com/Azure/azure-cosmos-dotnet-v3/pull/1832

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItem.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+///                 string rawPayload = encryptionException.EncryptedContent;
+///             }
+///         }
+///     }
+/// }
+/// ]]>
+/// </code>
+/// </example>
+
+// TODO: moderakh add example
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItem.java:49]: TODOs - are they tracked in a separate PR? or can this todo be removed due to the example provided above?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItemCore.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        ObjectNode decryptableContentAsObjectNode = Utils.as(this.decryptableContent, ObjectNode.class);
+        if (decryptableContentAsObjectNode == null) {
+            // (this.cosmosSerializer.FromStream<T>(EncryptionProcessor.BaseSerializer.ToStream
+            return Mono.just(new DecryptionResult<>(Utils.getSimpleObjectMapper().convertValue(this.decryptableContent, classType), null));
+        }
+
+        Mono<Pair<ObjectNode, DecryptionContext>> decryptedItemAndContextPairMono =
+            EncryptionProcessor.decrypt(decryptableContentAsObjectNode, this.encryptor);
+
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItemCore.java:48]: NIT remove on line

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ------------------------------------------------------------
+
+
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java:5]: NIT remove one line

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+///     //Handle and log exception
+///     return;
+/// }
+///
+/// (ToDoActivity toDo, DecryptionContext _) = await item.DecryptableItem.GetItemAsync<ToDoActivity>();
+/// ]]>
+/// </code>
+/// </example>
+
+// TODO moderakh add example
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java:49]: NIT remove todo

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java [hunk 5]
```diff
@@ -163,29 +204,66 @@
         EncryptionItemRequestOptions encryptionItemRequestOptions = Utils.as(requestOptions,
             EncryptionItemRequestOptions.class);
 
-        if (encryptionItemRequestOptions != null && encryptionItemRequestOptions.getEncryptionOptions() != null) {
-            Preconditions.checkArgument(partitionKey != null, "partitionKey cannot be null for operations using "
-                + "EncryptionContainer.");
+        if (encryptionItemRequestOptions == null || encryptionItemRequestOptions.getEncryptionOptions() == null) {
+            return container.createItem(item, partitionKey, requestOptions);
+        }
+
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java:210]: NIT remove one line - also below after the precondition check

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java [hunk 6]
```diff
@@ -163,29 +204,66 @@
...
+            return response
+                .publishOn(encryptionScheduler)
+                .map(
+                    rsp -> {
+                        return this.responseFactory.createItemResponse(rsp, (Class<T>) item.getClass());
+                    }
+                );
+
+            // TODO: no support for streaming
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java:257]: ???

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java [hunk 7]
```diff
@@ -261,48 +375,85 @@
...
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java:431]: Same as above - at least to me it isn't clear what this comment means. We don't support Stream Apis at all in Java,  correct?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java (hunk 6 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java:257]: we don't have support for streaming anywhere in the SDK.
I am considering wiring up some byte[] API. 

this TODO is only a hint to myself no actual work is remaining.
I will reword it to make it cleaner.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java (hunk 5 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java:210]: thanks. next commit will fix.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java (hunk 3 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java:5]: thanks next commit will fix.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItemCore.java (hunk 2 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItemCore.java:48]: thanks next commit will fix.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - I am not super happy with the naming EncrytableXXX vs. DecrytableXXX - but have tried to come up on the .Net PR already with a better name without much success. Otherwise except for checking which TODOs are still relevant, can be removed or should be formally tracked by GitHub items with reference to the link/item number in the source code everything looks fine and consistent with the .Net design

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItem.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/DecryptableItem.java:49]: the public APIs haven't finalized yet. For that reason I would like to leave wiring the examples for later (to avoid re-do work), once the final APIs are finalized.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java (hunk 4 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptableItem.java:49]: the public APIs haven't finalized yet. For that reason I would like to leave wiring the examples for later (to avoid re-do work), once the final APIs are finalized.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java (hunk 7 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/EncryptionCosmosAsyncContainer.java:431]: you are right the comment is confusing.
we don't support streaming api in java SDK.

will cleanup
